cmd/infra/powervs/destroy.go

```diff
@@ -0,0 +1,472 @@
+package powervs
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"k8s.io/apimachinery/pkg/util/errors"
+	"strings"
+	"time"
+
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/IBM-Cloud/power-go-client/clients/instance"
+	"github.com/IBM-Cloud/power-go-client/ibmpisession"
+	"github.com/IBM/platform-services-go-sdk/resourcecontrollerv2"
+	"github.com/IBM/vpc-go-sdk/vpcv1"
+)
+
+const (
+	// Time duration for monitoring the resource readiness
+	cloudInstanceDeletionTimeout   = time.Minute * 10
+	powerVSResourceDeletionTimeout = time.Minute * 5
+	vpcResourceDeletionTimeout     = time.Minute * 2
+	dhcpServerDeletionTimeout      = time.Minute * 10
+
+	// Resource desired states
+	powerVSCloudInstanceRemovedState = "removed"
+	powerVSJobCompletedState         = "completed"
+	powerVSJobFailedState            = "failed"
+	dhcpInstanceShutOffState         = "SHUTOFF"
+)
+
+type DestroyInfraOptions struct {
+	InfraID                    string
+	InfrastructureJson         string
+	ResourceGroup              string
+	PowerVSRegion              string
+	PowerVSZone                string
+	PowerVSCloudInstanceID     string
+	DeletePowerVSCloudInstance bool
+	PowerVSDhcpID              string
+	PowerVSCloudConnection     string
+	VpcRegion                  string
+	Vpc                        string
+	Debug                      bool
+}
+
+func NewDestroyCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:          "powervs",
+		Short:        "Destroys PowerVS infrastructure resources for a cluster",
+		SilenceUsage: true,
+	}
+
+	opts := DestroyInfraOptions{}
+
+	cmd.Flags().StringVar(&opts.InfraID, "infra-id", opts.InfraID, "Cluster ID with which to tag AWS resources (required)")
+	cmd.Flags().StringVar(&opts.InfrastructureJson, "infrastructure-json", opts.InfrastructureJson, "Result of ./hypershift infra create powervs")
+	cmd.Flags().StringVar(&opts.ResourceGroup, "resource-group", opts.ResourceGroup, "IBM Cloud Resource Group")
+	cmd.Flags().StringVar(&opts.VpcRegion, "vpc-region", opts.VpcRegion, "IBM Cloud VPC Infra Region")
+	cmd.Flags().StringVar(&opts.Vpc, "vpc", opts.Vpc, "IBM Cloud VPC")
+	cmd.Flags().StringVar(&opts.PowerVSRegion, "powervs-region", opts.PowerVSRegion, "PowerVS Region")
+	cmd.Flags().StringVar(&opts.PowerVSZone, "powervs-zone", opts.PowerVSZone, "IBM Cloud PowerVS Zone")
+	cmd.Flags().StringVar(&opts.PowerVSCloudConnection, "powervs-cloud-connection", opts.PowerVSCloudConnection, "IBM Cloud PowerVS Cloud Connection")
+	cmd.Flags().StringVar(&opts.PowerVSCloudInstanceID, "powervs-cloud-instance-id", opts.PowerVSCloudInstanceID, "IBM PowerVS Cloud Instance ID")
```

Comment:
why do we need all this flags? isn't the infraID and region enough?

Reply:
Need to get zone information along with region to create the resources.
powervs-cloud-connection, powervs-cloud-instance-id & vpc these flags are for dev and test purpose to use the existing resources instead of creating a new one. We have hidden these flags for the common usage.

Reply:
> to create the resources.

but this is destroy, right? If I created the infra with the cli I should be able to destroy with zone/region/infraID, I don't create the infra with cli I should not be able to destroy it with the cli. Does that make sense? So still not clear to me why we need those flags in destroy.

Reply:
We will setup few things like adding a DHCP server into the powervs-cloud-instance-id or attaching the vpc and DHCP private network to powervs-cloud-connection for setting up the hypershift infra. Those things we want to clean up, so that user needs to pass these ids so that destroy, and will clean up the resources created by the hypershift infra creation flow.